### PR TITLE
(WIP) doc-src: add preliminary openEMS documentation

### PR DESCRIPTION
**Pull request by biergaizi**
## WIP, do not merge, but comments welcomed

The following draft documentation adds a new section to the openEMS documentation website called "concepts". It explains several important concepts of CSXCAD, openEMS and FDTD. This represents a small part of my month-long work made in 2025 on documenting openEMS.

Currently this documentation is still incomplete, some parts contain known mistakes. To my best knowledge, I am aware of two mistakes and several limitations:

1. [X] "primitives.rst" contains many broken Python syntax in the example code - All fixed!

2. [ ] "ports.rst" does not correctly explain the use of transmission line ports, the given information is only correct for lumped port.

3. [ ] Post-processing is unexplained. This part needs to be "ported" from my parallel-plate waveguide tutorial.

4. [ ] No tutorial is given, my parallel-plate waveguide tutorial needs some rework.

I plan to fix (1) and (2) in the upcoming days, and get these documents merged. The rest can be added in the future. "Some documents" is always better than "no document".

Although it's still a work in progress, I believe this documentation draft already contains information not documented anywhere about the openEMS project. In particular, "signals.rst" explains the usage of all excitation waveforms, and "dispersive.rst" explains the mathematical formulation of Debye, Drude, Lorentz models, including comparison to other textbook formulations. Both are original work.

Comments (other than repeating the known problems) are welcomed since I may have still missed something.

### CODE REVIEW

File: doc-src/concepts/bc.rst [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+   A simulation box with PML at all boundaries acts like an anechoic chamber
+   in an EMC test lab, mimicking an infinitely large free space. Image by
+   Adamantios (at English Wikipedia), licensed under CC BY-SA 3.0.
+
+Two kinds of Absorbing Boundary Conditions are implemented in openEMS.
+
+#. **Mur's Boundary Condition (MUR)**. This is a
+   first-generation boundary condition purely defined by differential
+   equations, originally invented by Gerrit Mur in the 1980s. It has
+   a moderate computational overhead, but it works only if the EM wave
```
Comment (thliebig) [inline doc-src/concepts/bc.rst]: it works only at a orthogonal direction is not really correct. It works best (even perfect) for orthogonal direction, but it still works quite decent for other angles too. It just gets worse and worse the further away from orthogonal you get...

**Review by thliebig — COMMENTED** [review]
Comments for the first few files. This is going to take a while...

File: doc-src/concepts/bc.rst [hunk 2]
```diff
@@ -0,0 +1,180 @@
+Boundary Conditions
+=====================
+
+In openEMS, all physical phenomena occur in a small simulation box, so
```
[the remark this answers is not in the document]
Comment (biergaizi) [reply doc-src/concepts/bc.rst]: Right. s/in openEMS/in computational electromagnetism/.

**Review by thliebig — COMMENTED** [review]
Still not done, but only three *.rst to do...

@biergaizi please not that some comments may be hidden under the "Load diff"

File: doc-src/concepts/bc.rst [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+   second-generation boundary condition proposed in the 1990s, modeling
+   the behavior of a hypothetical EM wave-absorbing material. Unlike
+   Mur's ABC, PML occupies some physical cells in the simulation box
+   (the actual boundary at the true edge remains PEC).
+   This mimics the wave-absorbing foam on the wall of an anechoic chamber
+   in EMC test labs.
+
+   PML is a more effective absorber and is easy to use, but it has the
+   highest computational overhead (especially in openEMS, due to suboptimal
+   implementation). Avoid it if efficiency is critical (e.g. only use
```
[the remark this answers is not in the document]
Comment (thliebig) [reply doc-src/concepts/bc.rst]: Not sure if I wrote that or not. But in any case, it is as sub-optimal as any other extension. Since it tries to make use of the main engine operator maybe even less than others.
In any case, it is not interesting for the reader of the docu? Because if you need to use the PML you simply need to :smirk:

File: doc-src/concepts/mesh.rst [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+   1. Only the electric field is known at an exact mesh line. The magnetic field
+   cell is known at a position with a half-step offset (``x + 0.5``). It's
+   located in the middle of the two mesh lines.  All mesh lines, coordinates, and
+   CSXCAD visualization are based on the primary mesh. Current probes placed at
+   a mesh line is automatically snapped to the nearest secondary mesh line. Thus,
+   naive current probe placements create a slight error due to this misalignment.
+
+   2. Only the electric field is known at an exact timestep, the magnetic field
+   is known with a half-timestep offset (``t + 0.5``). This error is always
+   neglected.
```
[the remark this answers is not in the document]
Comment (thliebig) [reply doc-src/concepts/mesh.rst]: There is no correct time? Both fields are recorded with their respective time and always has to be taken care of properly.  Putting E- and H-fields in a direct relationship in time-domain is difficult and should be avoided I guess.

File: doc-src/development/python_todo.rst [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    and manually set the model parameters via
+    :meth:`~CSXCAD.CSProperties.CSProperties.SetAttributeValue`.
+
+  - For :func:`AddDjordjevicSarkarMaterial`, :func:`CalcDebyeMaterial`,
+    :func:`CalcDjordjevicSarkarApprox`, :func:`CalcDrudeMaterial`,
+    :func:`CalcLorentzMaterial`, no workaround is available. These
+    Octave functions are helper functions that calculate the model's
+    output curves for the purpose of fitting parameters and preparing
+    a simulation. No C++ APIs exist because they're not actually used
+    in the simulation.
```
[the remark this answers is not in the document]
Comment (thliebig) [reply doc-src/development/python_todo.rst]: Well this is a valid point about the policy. I think my current personal policy is, if an LLM is used e.g. to translate some function, but the author of that code than does a good review and testing of that code and is open about the whole process in the PR, this should be okay?? But I guess we should have a look at what LLM usage policy we want to adopt at some point (soon)...

**Review by thliebig — COMMENTED** [review]
Some comments on your comments

File: doc-src/concepts/mesh.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (biergaizi) [reply doc-src/concepts/mesh.rst]: This is consistent with my own understanding, the field dump of E and H are off by a half-timestep, the openEMS engine does not auto-interpolate two H fields at H(t - 0.5),  H(t + 0.5) to approximate H(t) - it's impossible even if someone wants to, because the engine only saves one copy of the field at the latest timestep, there's no double buffering.

This is why I said: "this error is always neglected", I meant it's neglected in openEMS engine's probes and field dumps.

I propose the following wording to clarify the issues. 

> Only the electric field is known at an exact timestep, the magnetic field is known with a half-timestep offset (``t + 0.5``). When probing both voltages and currents, or dumping both electric and magnetic fields, users must explicitly take this simulation artifact into account. Since temporal interpolation is unsupported, analyzing both physical quantities at the same timestep is difficult, and best be avoided without a strong FDTD background.

[the remark this answers is not in the document]
Comment (thliebig) [reply doc-src/concepts/mesh.rst]: We could add that this only affects the fields in time-domain? None of this issue arise if you are looking (at the more common?) frequency domain.

**Comment by VolkerMuehlhaus** [comment]
I don't know know how to create a link here, but the image for extruded polygons is wrong. This shows the image for rotated polygons.

https://openems--417.org.readthedocs.build/en/417/concepts/primitives.html

**Comment by biergaizi** [comment]
@VolkerMuehlhaus Fixed.

File: doc-src/concepts/dump.rst [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+However, some special
+applications make use of the raw electromagnetic fields, not just the
+input and output signals. We can do this by creating a "dump box" (a region
+in space where field values are recorded) to save field samples to disk.
+For troubleshooting malfunctioning setups, this is especially helpful as one can
+identify the problematic region through direct visualization.
+
+Several kinds of dump boxes exist.
+
+#. Time-domain dumps of electric field :math:`\mathbf{E}`, auxiliary magnetic
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply doc-src/concepts/dump.rst:18]: Also compare the first and second paragraph in [this section](https://meep.readthedocs.io/en/master/Introduction/#maxwells-equations) of the MEEP documentation:

> Meep simulates [Maxwell's equations](https://en.wikipedia.org/wiki/Maxwell's_equations), which describe the interactions of electric (E) and magnetic (H) fields

> where D is the displacement field, ε is the dielectric constant, J is the current density (of electric charge), and JB is the magnetic-charge current density. Magnetic currents are a convenient computational fiction in some situations. **B is the magnetic flux density (often called the magnetic field)**, μ is the magnetic permeability, and H is the magnetic field.

**Comment by thliebig** [comment]
I have taken the current draft and cleaned it up a bit etc. 
I will shorty squash merge this.
Thanks to everyone that has contributed. 
I think I will close this draft PR? Ideally for new additional topics a separate PR is better to handle?
I can always re-open if someone disagrees and has a good point ;)

**Comment by thliebig** [comment]
The squash-merged commit is: thliebig/openEMS-Project@352096848184d606a0c8e5d15a78ac940fbef98e
